File: flatten_a_multilevel_doubly_linked_list/FlattenAMultilevelDoublyLinkedList.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node prev;
    public Node next;
    public Node child;
};
*/

class FlattenAMultilevelDoublyLinkedList {
    public Node flatten(Node head) {
        if (head == null) {
            return head;
        }
        if (head.child == null) {
            flatten(head.next);
            return head;
        }

        Node childNode = head.child;
        Node nextNode = head.next;
        head.child = null;
        if (nextNode != null) {
            nextNode.prev = null;
        }

        head.next = flatten(childNode);
        head.next.prev = head;

        Node curr = head;
        while (curr != null && curr.next != null) {
            curr = curr.next;
        }
        curr.next = nextNode;
        if (nextNode != null) {
            curr.next.prev = curr;
        }

        return head;
    }
}

File: flatten_a_multilevel_doubly_linked_list/Flatten_A_Multilevel_Doubly_Linked_List.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};
*/

class Flatten_A_Multilevel_Doubly_Linked_List {
public:
    Node* flatten(Node* head) {
        if (head == NULL) {
            return head;
        }
        if (head->child == NULL) {
            flatten(head->next);
            return head;
        }

        Node* child = head->child;
        Node* next = head->next;
        head->child = NULL;
        head->next = NULL;
        if (next != NULL) {
            next->prev = NULL;
        }

        head->next = flatten(child);
        head->next->prev = head;

        Node* curr = head;
        while (curr != NULL && curr->next != NULL) {
            curr = curr->next;
        }
        curr->next = next;
        if (next != NULL) {
            next->prev = curr;
        }

        return head;
    }
};